src/core/dater.rs

```diff
@@ -0,0 +1,209 @@
+    fn new_with_dtsb(dts: &[u8]) -> Result<Matter>;
+    fn new_with_qb64(qb64: &str) -> Result<Matter>;
+    fn new_with_qb64b(qb64b: &[u8]) -> Result<Matter>;
+    fn new_with_qb2(qb2: &[u8]) -> Result<Matter>;
+    fn dts(&self) -> Result<String>;
+    fn dtsb(&self) -> Result<Vec<u8>>;
+}
+
+fn validate_code(code: &str) -> Result<()> {
+    lazy_static! {
+        static ref CODES: Vec<&'static str> = vec![matter::Codex::DateTime.code(),];
+    }
+
+    if !CODES.contains(&code) {
+        return err!(Error::UnexpectedCode(code.to_string()));
+    }
+
+    Ok(())
+}
+
+fn iso_8601_to_b64(dts: &str) -> String {
+    dts.replace(':', "c").replace('.', "d").replace('+', "p")
+}
+
+fn b64_to_iso_8601(dts: &str) -> String {
+    dts.replace('c', ":").replace('d', ".").replace('p', "+")
+}
+
+fn now_as_iso8601() -> String {
+    let dt = chrono::offset::Utc::now();
+    dt.to_rfc3339_opts(chrono::SecondsFormat::Micros, false)
+}
+
+fn now_as_b64() -> String {
+    let dt = now_as_iso8601();
+    iso_8601_to_b64(&dt)
+}
+
+impl Dater for Matter {
+    fn new_with_code_and_raw(code: &str, raw: &[u8]) -> Result<Matter> {
+        if !code.is_empty() {
+            validate_code(code)?;
+        }
+        if raw.is_empty() {
+            let qb64 = format!("{}{}", matter::Codex::DateTime.code(), now_as_b64());
+            Matter::new_with_qb64(&qb64)
+        } else {
+            Matter::new_with_code_and_raw(code, raw)
+        }
+    }
+
+    fn new_with_dts(dts: &str) -> Result<Matter> {
+        let b64 = if dts.is_empty() { now_as_b64() } else { iso_8601_to_b64(dts) };
+        let qb64 = format!("{}{}", matter::Codex::DateTime.code(), &b64);
+        Matter::new_with_qb64(&qb64)
+    }
+
+    fn new_with_dtsb(dts: &[u8]) -> Result<Matter> {
+        Matter::new_with_dts(&String::from_utf8(dts.to_vec())?)
+    }
+
+    fn new_with_qb64(qb64: &str) -> Result<Matter> {
+        let dater = Matter::new_with_qb64(qb64)?;
+        validate_code(&dater.code)?;
+        Ok(dater)
+    }
+
+    fn new_with_qb64b(qb64b: &[u8]) -> Result<Matter> {
+        let dater = Matter::new_with_qb64b(qb64b)?;
+        validate_code(&dater.code)?;
+        Ok(dater)
+    }
+
+    fn new_with_qb2(qb2: &[u8]) -> Result<Matter> {
+        let dater = Matter::new_with_qb2(qb2)?;
+        validate_code(&dater.code)?;
+        Ok(dater)
+    }
+
+    fn dts(&self) -> Result<String> {
+        let hs = matter::sizage(&self.code)?.hs as usize;
+        let qb64 = self.clone().qb64()?;
+        Ok(b64_to_iso_8601(&qb64[hs..]))
+    }
+
+    fn dtsb(&self) -> Result<Vec<u8>> {
+        Ok(self.dts()?.as_bytes().to_vec())
+    }
+}
+
+#[cfg(test)]
+mod test_dater {
+    use super::{matter, Dater, Matter};
+    use rstest::rstest;
+
+    #[rstest]
+    fn test_new_default(
+        #[values(
+        &<Matter as Dater>::new_with_code_and_raw("", &[]).unwrap(),
+        &<Matter as Dater>::new_with_dts("").unwrap(),
+        &<Matter as Dater>::new_with_dtsb(b"").unwrap(),
```

Comment:
Is this the equivalent of pytest's parametrize? https://docs.pytest.org/en/6.2.x/parametrize.html

If it is, amazing.

Reply:
Yeah based on what I see it must be. rstest ftw.

Reply:
More or less. At least it's very easy and allows to build matrices multiplying `values` vectors. It's awesome.